File: examples/rs/06_graphdata_large_payload_arc_or_shared_data.rs
// Example 06: GraphData with Arc (Large Payload Sharing)
// Demonstrates efficient memory sharing for large data using Arc

mod benchmark_utils;

use dagex::{Graph, GraphData};
use std::collections::HashMap;
use benchmark_utils::{Benchmark, print_header, print_section};

fn create_large_data(_inputs: &HashMap<String, GraphData>) -> HashMap<String, GraphData> {
    // Create a large vector (simulating a large dataset)
    let large_vec: Vec<i64> = (0..1_000_000).collect();
    
    let mut outputs = HashMap::new();
    // Store using Arc (int_vec automatically wraps in Arc)
    outputs.insert("large_data".to_string(), GraphData::int_vec(large_vec));
    outputs
}

fn consumer_a(inputs: &HashMap<String, GraphData>) -> HashMap<String, GraphData> {
    let mut outputs = HashMap::new();
    
    if let Some(data) = inputs.get("data") {
        if let Some(vec) = data.as_int_vec() {
            // Access the data through Arc - no copying!
            let sum: i64 = vec.iter().take(1000).sum();
            outputs.insert("sum_a".to_string(), GraphData::int(sum));
        }
    }
    
    outputs
}

fn consumer_b(inputs: &HashMap<String, GraphData>) -> HashMap<String, GraphData> {
    let mut outputs = HashMap::new();
    
    if let Some(data) = inputs.get("data") {
        if let Some(vec) = data.as_int_vec() {
            // Access the data through Arc - no copying!
            let sum: i64 = vec.iter().skip(1000).take(1000).sum();
            outputs.insert("sum_b".to_string(), GraphData::int(sum));
        }
    }
    
    outputs
}

fn consumer_c(inputs: &HashMap<String, GraphData>) -> HashMap<String, GraphData> {
    let mut outputs = HashMap::new();
    
    if let Some(data) = inputs.get("data") {
        if let Some(vec) = data.as_int_vec() {
            // Access the data through Arc - no copying!
            let sum: i64 = vec.iter().skip(2000).take(1000).sum();
            outputs.insert("sum_c".to_string(), GraphData::int(sum));
        }
    }
    
    outputs
}

fn main() {
    print_header("Example 06: GraphData with Arc (Large Payload Sharing)");
    
    println!("📖 Story:");
    println!("   When working with large data, copying it between nodes is expensive.");
    println!("   GraphData automatically wraps large vectors (int_vec, float_vec) in Arc");
    println!("   for zero-copy sharing. Multiple nodes can read the same data without");
    println!("   duplication, saving both time and memory.\n");
    
    print_section("Building the Graph");
    
    let mut graph = Graph::new();
    
    // Create large data source
    graph.add(
        create_large_data,
        Some("CreateLargeData"),
        None,
        Some(vec![("large_data", "data")])
    );
    
    // Add multiple consumers that share the large data
    graph.add(
        consumer_a,
        Some("ConsumerA"),
        Some(vec![("data", "data")]),
        Some(vec![("sum_a", "sum_a")])
    );
    
    graph.add(
        consumer_b,
        Some("ConsumerB"),
        Some(vec![("data", "data")]),
        Some(vec![("sum_b", "sum_b")])
    );
    
    graph.add(
        consumer_c,
        Some("ConsumerC"),
        Some(vec![("data", "data")]),
        Some(vec![("sum_c", "sum_c")])
    );
    
    let dag = graph.build();
    
    print_section("Mermaid Diagram");
    println!("{}", dag.to_mermaid());
    
    print_section("ASCII Visualization");
    println!("                      ConsumerA");
    println!("                     /");
    println!("  CreateLargeData  -- ConsumerB");
    println!("   (1M integers)    \\");
    println!("                      ConsumerC");
    
    println!("💡 Key insight: The large data (1M integers) is created once");
    println!("   and shared via Arc. No copies are made when distributing");
    println!("   to multiple consumers.\n");
    
    print_section("Sequential Execution (parallel=false)");
    
    let bench = Benchmark::start("Sequential execution");
    let _context_seq = dag.execute(false, None);
    let result_seq = bench.finish_and_print();
    
    print_section("Parallel Execution (parallel=true)");
    
    let bench = Benchmark::start("Parallel execution");
    let context_par = dag.execute(true, Some(4));
    let result_par = bench.finish_and_print();
    
    print_section("Results");
    
    println!("📊 Consumer outputs (each processes different segments):");
    
    if let Some(sum_a) = context_par.get("sum_a").and_then(|d| d.as_int()) {
        println!("   ConsumerA (first 1000):  sum = {}", sum_a);
    }
    
    if let Some(sum_b) = context_par.get("sum_b").and_then(|d| d.as_int()) {
        println!("   ConsumerB (next 1000):   sum = {}", sum_b);
    }
    
    if let Some(sum_c) = context_par.get("sum_c").and_then(|d| d.as_int()) {
        println!("   ConsumerC (next 1000):   sum = {}", sum_c);
    }
    
    println!("\nSequential execution:");
    println!("  Time: {:.3}ms", result_seq.duration_ms);
    
    println!("\nParallel execution:");
    println!("  Time: {:.3}ms", result_par.duration_ms);
    
    println!("\n✅ Zero-copy data sharing successful!");
    println!("   Memory benefit: Only 1 copy of data exists, shared by all consumers");
    
    println!();
}

File: examples/py/06_graphdata_large_payload_arc_or_shared_data.py
"""Example 06: GraphData with Shared Data (Large Payload Sharing)

Demonstrates efficient memory handling for large data in Python
"""

import sys
sys.path.insert(0, '/home/runner/work/graph-sp/graph-sp/examples/py')

from benchmark_utils import Benchmark, print_header, print_section
import dagex


def create_large_data(_inputs):
    """Create a large dataset."""
    # Create a large list (simulating a large dataset)
    large_list = list(range(1_000_000))
    
    # In Python, objects are reference-counted, so passing the list
    # to multiple consumers doesn't create copies unless modified
    return {"large_data": large_list}


def consumer_a(inputs):
    """Consumer A: process first segment."""
    data = inputs.get("data", [])
    # Access the data by reference - Python won't copy unless we modify
    sum_a = sum(data[:1000])
    return {"sum_a": sum_a}


def consumer_b(inputs):
    """Consumer B: process second segment."""
    data = inputs.get("data", [])
    # Access the data by reference - Python won't copy unless we modify
    sum_b = sum(data[1000:2000])
    return {"sum_b": sum_b}


def consumer_c(inputs):
    """Consumer C: process third segment."""
    data = inputs.get("data", [])
    # Access the data by reference - Python won't copy unless we modify
    sum_c = sum(data[2000:3000])
    return {"sum_c": sum_c}


def main():
    print_header("Example 06: GraphData with Shared Data (Large Payload)")
    
    print("📖 Story:")
    print("   When working with large data, copying it between nodes is expensive.")
    print("   In Python, objects are reference-counted, so passing lists/arrays")
    print("   to multiple consumers shares references rather than copying data.")
    print("   For even better performance, use NumPy arrays which are efficiently")
    print("   passed through the Python/Rust boundary.\n")
    
    print_section("Building the Graph")
    
    graph = dagex.Graph()
    
    # Create large data source
    graph.add(
        create_large_data,
        label="CreateLargeData",
        inputs=None,
        outputs=[("large_data", "data")]
    )
    
    # Add multiple consumers that share the large data
    graph.add(
        consumer_a,
        label="ConsumerA",
        inputs=[("data", "data")],
        outputs=[("sum_a", "sum_a")]
    )
    
    graph.add(
        consumer_b,
        label="ConsumerB",
        inputs=[("data", "data")],
        outputs=[("sum_b", "sum_b")]
    )
    
    graph.add(
        consumer_c,
        label="ConsumerC",
        inputs=[("data", "data")],
        outputs=[("sum_c", "sum_c")]
    )
    
    dag = graph.build()
    
    print_section("Mermaid Diagram")
    print(dag.to_mermaid())
    
    print_section("ASCII Visualization")
    print("                      ConsumerA")
    print("                     /")
    print("  CreateLargeData  -- ConsumerB")
    print("   (1M integers)    \\")
    print("                      ConsumerC")
    
    print("💡 Key insight: The large data (1M integers) is created once")
    print("   and shared by reference. Python's reference counting ensures")
    print("   efficient memory usage.\n")
    
    print_section("Sequential Execution (parallel=False)")
    
    with Benchmark("Sequential execution") as bench_seq:
        context_seq = dag.execute(parallel=False)
    
    bench_seq.print_result()
    result_seq = bench_seq.result
    
    print_section("Parallel Execution (parallel=True)")
    
    with Benchmark("Parallel execution") as bench_par:
        context_par = dag.execute(parallel=True, max_threads=4)
    
    bench_par.print_result()
    result_par = bench_par.result
    
    print_section("Results")
    
    print("📊 Consumer outputs (each processes different segments):")
    
    sum_a = context_par.get("sum_a")
    if sum_a is not None:
        print(f"   ConsumerA (first 1000):  sum = {sum_a}")
    
    sum_b = context_par.get("sum_b")
    if sum_b is not None:
        print(f"   ConsumerB (next 1000):   sum = {sum_b}")
    
    sum_c = context_par.get("sum_c")
    if sum_c is not None:
        print(f"   ConsumerC (next 1000):   sum = {sum_c}")
    
    print("\nSequential execution:")
    print(f"  Time: {result_seq.duration_ms:.3f}ms")
    
    print("\nParallel execution:")
    print(f"  Time: {result_par.duration_ms:.3f}ms")
    
    print("\n✅ Reference-based data sharing successful!")
    print("   Memory benefit: Data shared by reference, not copied")
    
    print()


if __name__ == "__main__":
    main()
